protocol/collectd/collectd.go

```diff
@@ -75,7 +75,13 @@ func NewDatapoint(point *JSONWriteFormat, index uint, defaultDimensions map[stri
 	addIfNotNullOrEmpty(dimensions, "dsname", !usedDsName, dsname)
 
 	timestamp := time.Unix(0, int64(float64(time.Second)**point.Time))
-	return datapoint.New(metricName, dimensions, datapoint.NewFloatValue(*val), metricType, timestamp)
+	var value datapoint.Value
+	if *val == float64(int64(*val)) {
```

Comment:
You might save a bit by keeping around the int value (although that probably gets optimized?):

```go
intVal := int64(*val)
if *val == float64(intVal) {
    value = datapoint.NewIntValue(intVal)
} else {
    value = datapoint.newFloatValue(*val)
}
```